jobfunnel/backend/tools/delay.py

```diff
@@ -0,0 +1,117 @@
+    # any delay of .2 or less is hard delay
+    if its <= 1:
+        return _c_delay(list_len, delay)
+    else:
+        # prevents slicing from breaking if delay is a float
+        if isinstance(its, float):
+            its = int(ceil(its))
+        # create list of x values based on scrape list size
+        delays = [*range(list_len)]
+        delays[0:its] = [x / 5 for x in delays[0:its]]
+        delays[its:] = [delay] * (len(delays) - its)
+        return delays
+
+
+def _sig_delay(list_len: int, delay: Union[int, float]):
+    """Calculates Richards/Sigmoid curve for delay.
+    NOTE: https://en.wikipedia.org/wiki/Generalised_logistic_function
+    """
+    gr = sqrt(delay) * 4  # growth rate
+    y_0 = log(4 * delay)  # Y(0)
+    # calculates sigmoid curve using vars rewritten to be our x
+    delays = delay * expit(arange(list_len) / gr - y_0)
+    return delays.tolist()  # convert np array back to list
+
+
+def calculate_delays(list_len: int, delay_config: DelayConfig) -> List[float]:
+    """Checks delay config and returns calculated delay list.
+
+    NOTE: we do this to be respectful to online job sources
+    TODO: we should be able to calculate delays on-demand.
+
+    Args:
+        list_len: length of scrape job list
+        delay_config: Delay configuration dictionary
+
+    Returns:
+        list of delay time matching length of scrape job list
+    """
+    delay_config.validate()
+
+    # Delay calculations using specified equations
+    if delay_config.algorithm == DelayAlgorithm.CONSTANT:
+        delay_vals = _c_delay(list_len, delay_config.max_duration)
+    elif delay_config.algorithm == DelayAlgorithm.LINEAR:
+        delay_vals = _lin_delay(list_len, delay_config.max_duration)
+    elif delay_config.algorithm == DelayAlgorithm.SIGMOID:
+        delay_vals = _sig_delay(list_len, delay_config.max_duration)
+    else:
+        raise ValueError(f"Cannot calculate delay for {delay_config.algorithm}")
+
+    # Check if minimum delay is above 0 and less than last element
+    if 0 < delay_config.min_duration:
+        # sets min_duration to values greater than itself in delay_vals
+        for i, n in enumerate(delay_vals):
+            if n > delay_config.min_duration:
+                break
+            delay_vals[i] = delay_config.min_duration
+
+    # Outputs final list of delays rounded up to 3 decimal places
+    if delay_config.random:  # check if random delay was specified
+        # random.uniform(a, b) a = lower bound, b = upper bound
+        if delay_config.converge:  # checks if converging delay is True
+            # delay_vals = lower bound, delay = upper bound
+            durations = [
+                round(uniform(x, delay_config.max_duration), 3)
+                for x in delay_vals
+            ]
+        else:
+            # lb = lower bounds, delay_vals = upper bound
+            durations = [
+                round(uniform(delay_config.min_delay, x), 3) for x in delay_vals
```

Comment:
_delay_config.min_delay_ should be changed to _delay_config.min_duration_.

Reply:
fixed